File: backend/app/services/audit.py
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.audit_log import AuditLog
from app.models.user import User


async def log_action(
    db: AsyncSession,
    *,
    user: User,
    action: str,
    module: str,
    entity_type: str | None = None,
    entity_id: int | None = None,
    description: str | None = None,
    ip_address: str | None = None,
    old_values: dict | None = None,
    new_values: dict | None = None,
    authorized_by: User | None = None,
    pin_verified: bool | None = None,
) -> AuditLog:
    log = AuditLog(
        user_id=user.id,
        user_email=user.email,
        action=action,
        module=module,
        entity_type=entity_type,
        entity_id=entity_id,
        description=description,
        company_id=user.company_id,
        ip_address=ip_address,
        old_values=old_values,
        new_values=new_values,
        authorized_by_user_id=authorized_by.id if authorized_by else None,
        authorized_by_email=authorized_by.email if authorized_by else None,
        pin_verified=pin_verified,
    )
    db.add(log)
    return log

File: frontend/src/types/audit.ts
export interface AuditLog {
  id: number;
  user_id: number | null;
  user_email: string;
  action: string;
  module: string;
  entity_type: string | null;
  entity_id: number | null;
  description: string | null;
  company_id: number | null;
  ip_address: string | null;
  authorized_by_user_id: number | null;
  authorized_by_email: string | null;
  pin_verified: boolean | null;
  old_values: Record<string, unknown> | null;
  new_values: Record<string, unknown> | null;
  timestamp: string;
}
